logger: Add a Segger RTT backend

pkral78 opened the pull request:
Add logger backend that uses Segger RTT for message output. Several
options are provided allowing configuration of up-buffer and logger
behaviour.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 committed:
logger: Add a Segger RTT backend

Add logger backend that uses Segger RTT for message output. Several
options are provided allowing configuration of up-buffer and logger
behaviour.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 committed:
logger: fix grammar/punctuation in rtt backend Kconfig

Fix some czenglish grammar issues.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

jhedberg commented:
How does this relate to #9365

pkral78 commented:
@jhedberg In fact I overlook that somebody already open this kind of PR (despite fact I have search for it). However looking and #9365  I hope that this PR is more complete and allow to choose blocking/non-blocking behaviour.

pkral78 committed:
ext: debug: segger: Terminal sharing using SEGGER_RTT_LOCK/UNLOCK macros

This commit allows to share default RTT (terminal) buffer with various
Zephyr subsystems.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 commented:
@jhedberg 
#9365 fixed by commit https://github.com/zephyrproject-rtos/zephyr/pull/10659/commits/e1f3fc87198d43a91fa6bd42c9599ca729641570 in this PR.

pkral78 committed:
ext: debug: segger: Add USE_SEGGER_RTT that enables RTT libraries.

This option allows to explicitly include Segger RTT libraries and
enables use of it for various subsystems. It is disabled by
default as it consumes more RAM.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

dbkinder commented on subsys/logging/Kconfig:
```diff
@@ -288,6 +288,66 @@ config LOG_BACKEND_UART
+	range 0 SEGGER_RTT_MAX_NUM_UP_BUFFERS
+	default 0
+	help
+	  Select index of up-buffer used for logger output, by default is uses
```
```suggestion
	  Select index of up-buffer used for logger output, by default it uses
```

nordic-krch commented on subsys/logging/log_backend_rtt.c:
```diff
@@ -0,0 +1,220 @@
+	return (int) (pos - data);
+}
+
+static int line_out(u8_t data)
```
could you explain bit more complexity of writing to RTT buffer. Seems to me that you could use log_output module in a better way, by providing bigger buffer instead of 1 byte (`LOG_OUTPUT_DEFINE(log_output, msg_out, &char_buf, 1)`). Then msg_out would be called only when this buffer is full or at the end of log message with the remainder.

pkral78 replied:
Well this is all about how out_func in log_output.c behave, it will not call log_output_flush until output buffer is full. I'd like to send log message to host as soon as EOL is reached without futher delays.

nordic-krch commented on subsys/logging/log_backend_rtt.c:
```diff
@@ -0,0 +1,220 @@
+
+#if CONFIG_LOG_BACKEND_RTT_BUFFER > 0
+
+#define RTT_LOCK()
```
please comment why when BUFFER>0 then lock is not needed.

pkral78 replied:
Because rtt backend then use differend non-shared memory region.

nordic-krch commented:
My general comments is that line approach with dropped message is complex and has some limitations:
- more ram used since additional line buffer is needed
- byte by byte processing 
- trimming when log message is longer than line_buf
In my opinion, that will cause more issues than the gain of linewise processing. I would use log_output buffer and if RTT cannot accept it we can inject `~` character indicating that something is lost. What do we gain by knowing how many lines got lost? If buffer is too small I will see some lines misformated and some `~`. That will be enough for me to increase the buffer or disable some  logs.

Nice that you've addressed #9778.

nordic-krch commented on subsys/logging/log_backend_rtt.c:
```diff
@@ -0,0 +1,220 @@
+	return 0;
+}
+
+#if CONFIG_LOG_BACKEND_RTT_MODE_DROP
```
if possible try to avoiding #if you can do something like:
```
int log_backend_rtt_write(void)
{
    if (IS_ENABLED(CONFIG_LOG_BACKEND_RTT_MODE_DROP)) { drop_write()}
   else block_write()
}
```

Then compiler will cut what is not used but will compile everything. As block approach is the default one, i'm affraid that at some point drop will not compile because it won't be that frequently used (e.g. segger api changes).

pkral78 replied:
NP, sounds reasonable, will do, however in this case intersection of both block uses same set of SEGGER API.
Do you still want for me to change that?

nordic-krch replied:
Segger API change was only the example. It's rather about the approach and not that particular case. The thing is that if you use IS_ENABLED macro you get all your code compiled even for not used options.

pkral78 replied:
Agree, better safe than sorry. Will do.

pkral78 replied:
Fixed by https://github.com/zephyrproject-rtos/zephyr/pull/10659/commits/06b91b792826215048f7166a032be0fab4d71a43

pkral78 commented:
> My general comments is that line approach with dropped message is complex and has some limitations:
> 
> * more ram used since additional line buffer is needed
> * byte by byte processing
> * trimming when log message is longer than line_buf

Processing line by line need additional buffer yes (about 128 bytes, what's the deal?), however if we are talking about sharing Terminal buffer with other subsystems as console, it is definitely better to send message whole at once. Then line buffer is need anyway.

Option for dropping messages is for cases where there is expected lot of debug output that can not be hold in memory. It can be turn off (which is default settings).

>   In my opinion, that will cause more issues than the gain of linewise processing. I would use log_output buffer and if RTT cannot accept it we can inject `~` character indicating that something is lost. What do we gain by knowing how many lines got lost? If buffer is too small I will see some lines misformated and some `~`. That will be enough for me to increase the buffer or disable some  logs.
> 

Well then you will see ~ for log_output buffers (not messages) that was rejected by RTT which I think print pretty mess. Also please consider sharing as I pointed before.


> Nice that you've addressed #9778.

NP.

dbkinder committed:
ext: debug: segger: Fix grammar in Kconfig

Fix grammar in ext/debug/segger/Kconfig

Signed-off-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

dbkinder committed:
subsys: logging: Fix grammar in Kconfig

Fix grammar in subsys/logging/Kconfig
Fix grammar in subsys/logging/Kconfig
Fix grammar in subsys/logging/Kconfig
Fix grammar in subsys/logging/Kconfig

Signed-off-by: David Kinder <david.b.kinder@intel.com>
Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 committed:
logger: fix PR codestyle issues

Fix some issues as complained by shippable.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 committed:
logger: better code maintainability in rtt backend

Better use of preprocessor to be aware of external library API
changes.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 commented:
One more thing ... what is purpose of HAS_SEGGER_RTT ? This symbol is derived from Kconfig on platforms which include nRF, Kinetis, LPC an iMX currently. However this give me no sense. Use of RTT  only depends on used JTAG/SWD probe. For example Atmel boards equipped with EGDB may be updated to J-Link FW and then are able to use RTT. Or boards can use external probe directly. Better to completely remove this and leave decision on application. @nordic-krch what you think?

carlescufi commented:
> One more thing ... what is purpose of HAS_SEGGER_RTT ? This symbol is derived from Kconfig on platforms which include nRF, Kinetis, LPC an iMX currently. However this give me no sense. Use of RTT only depends on used JTAG/SWD probe. For example Atmel boards equipped with EGDB may be updated to J-Link FW and then are able to use RTT. Or boards can use external probe directly. Better to completely remove this and leave decision on application. @nordic-krch what you think?

That is a good point. I added that option originally to signal in some way that the hardware was Segger-compatible, but with external J-Link probes or firmware updates this becomes a non-issue.

pkral78 commented:
> That is a good point. I added that option originally to signal in some way that the hardware was Segger-compatible, but with external J-Link probes or firmware updates this becomes a non-issue.

Then I could remove HAS_SEGGER_RTT and leave only USE_SEGGER_RTT or maybe better set HAS_SEGGER_RTT only for ARM platforms (and Renesas in future).

carlescufi commented:
>Then I could remove HAS_SEGGER_RTT and leave only USE_SEGGER_RTT or maybe better set HAS_SEGGER_RTT only for ARM platforms (and Renesas in future).

Fine but let’s do that in a follow-up PR once this is merged

dbkinder reviewed (approved):
LGTM, thanks.

pizi-nordic reviewed (commented):
I applogize, but I cannot finish my review now. I will try to post all my comments today.

pkral78 commented on ext/debug/segger/rtt/SEGGER_RTT_Conf.h:
```diff
@@ -138,7 +138,12 @@ else
 */
 #if (defined __SES_ARM) || (defined __CROSSWORKS_ARM) || (defined __GNUC__)
-  #ifdef __ARM_ARCH_6M__
+  #ifdef __ZEPHYR__
```
> What about calling Segger API from interrupt context?
Why we want to allow that ? If default buffers (aka Terminal) are shared across various Zephyr subsystems, we should just disallow that. So its up to caller how to dealt with that. 
- option 1 - caller may not use default buffer and allocate differend up/down buffers as they are not shared (they are not protected by mutex)
- option 2 - caller want to use default buffer, so he must use some queue and sync data from ISR and threads
> Please note, that logger could enter panic mode in which all log processing is synchronous.
This is handled, please read this PR as whole.
> This means, that LOG_XXX() could call Segger from any context, including fault handers etc.
This is handled, in panic mode it simply does not respect mutex and take buffers over.

pkral78 committed:
logger: fix line_pos setup in rtt backend drop mode

Fix initial write position in line_buf when drop mode is selected.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 commented on subsys/logging/log_backend_rtt.c:
```diff
@@ -0,0 +1,238 @@
+		/* pass */
+	}
+}
+
```
I think that log_backend_rtt_flush may be improved like this:

`
while(SEGGER_RTT_HasDataUp(CONFIG_LOG_BACKEND_RTT_BUFFER)) {
     if(!panic_mode) {
          k_yield();
     }
}
`

This ensure that other threads do not starve or depend on round robin scheduling. What do you think?

pizi-nordic commented on subsys/logging/log_backend_rtt.c:
```diff
@@ -0,0 +1,238 @@
+
+#if CONFIG_LOG_BACKEND_RTT_MODE_DROP
+
+#define DROP_MSG "\nmessages dropped: 99 \r"
```
Hardcoded number, related to other hardcoded number.
IMHO this 99 should be replaced by stringified define.

pkral78 committed:
logger: replace hardcoded constant in rtt backend

Replace hardcoded report limit of dropped message by macro.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 committed:
logger: console: init rtt only once

As RTT is used by more subsystem in Zephyr, initialization is moved
from RTT console and RTT logger backend to common kernel init phase.

Signed-off-by: Pavel Kral <pavel.kral@omsquare.com>

pkral78 commented:
Note that recent https://github.com/zephyrproject-rtos/zephyr/pull/10659/commits/31de2db76c200e826c07deeae9fdb0c6bb6d5d60 initializes RTT in kernel init phase.

carlescufi commented:
@nordic-krch @pizi-nordic can you please re-review?

nordic-krch reviewed (approved):
I'm not super happy with line wise approach (adds ram, byte processing where bulk processing could be applied) but it can be extended in the future to support that mode so i'm approving.

pkral78 commented:
@nordic-krch agree with that and that could be done by introducing some logger backend flags for finer grained log output control. But for current state this just works. Let's wait and see if there will be more requirements after introducing another logger backends.

Anyway thanks for all your comments guys.
